cmd/snap-update-ns/change.go

```diff
@@ -59,26 +60,141 @@ var changePerform func(*Change) ([]*Change, error)
 
 // changePerformImpl is the real implementation of Change.Perform
 func changePerformImpl(c *Change) ([]*Change, error) {
+	var changes []*Change
 	if c.Action == Mount {
+		// We may be asked to bind mount a file, bind mount a directory, mount
+		// a filesystem over a directory or create a symlink (which is abusing
+		// the "mount" concept slightly). That actual operation is performed in
+		// c.lowLevelPerform. Here we just set the stage to make that possible.
+		kind, _ := c.Entry.OptStr("x-snapd.kind")
+		path := c.Entry.Dir
+
+		// We use lstat to ensure that we don't follow a symlink in case one
+		// was set up by the snap. Note that at the time this is run, all the
+		// snap's processes are frozen.
+		fi, err := osLstat(path)
+
+		// In case we need to create something, some constants.
 		const (
 			mode = 0755
 			uid  = 0
 			gid  = 0
 		)
-		// Create target mount directory if needed.
-		if err := ensureMountPoint(c.Entry.Dir, mode, uid, gid); err != nil {
-			return nil, err
+
+		if err != nil && os.IsNotExist(err) {
+			// If the element doesn't exist we can attempt to create it.
+			// We will create the parent directory and then the final element
+			// relative to it. The traversed space may be writable so we try a
+			// to create the parent directory first.
+			switch kind {
+			case "":
```

Comment:
I find it slightly akward that we use "" to indicate directories, and not "dir" to be consistent. But I see this was already implemented before (and perhaps deliberately made that way), so just mentioning in case you have second thoughs and want to change it.

Reply:
Same here, is it possible to still change this? Or is this a backward compat problem? Maybe we could at least support both: `case "", "dir":` ?

Reply:
@mvo5 changing "" to "dir" is a backwards compatibility problem that I'd like to avoid. To unbreak it would require extra complexity in the diff calculation logic. If you feel strong about it I will do that in an iteration after this one.